### DISCUSSION TITLE
Allow methods to explicitly implement delegate-types

### DISCUSSION THREAD

**Discussion by DanielBiskup** [dotnet/csharplang#6425] [Language Ideas]
Using the [examples from the docs](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/delegates/using-delegates), let's say that you have a delegate-type `Del` defined as:
```csharp
public delegate void Del(string message);
```
To me delegates are like interfaces but for functions. They define what signature a function should implement. 
Now what I'd like to be able to do is to define a method that **explicitly** implements the delegate-type `Del`, i.e. a method that has the same signature as defined by `Del`.

I'll explain what I mean by explicit and give an example of why I this might be useful.
If you have a class `A` that provides a public `event` of type `Del` and a class `B` that wants to subscribe to it. That would roughly look as follows:
```csharp
public delegate void Del(string message);

public static class X {
  public static void SomeMethod(string message) { }
}

public class A {
  public event Del SomeEvent;
  public void FireEvent() { SomeEvent?.Invoke("Some Message");
}

public class B {
  private A _a;
  void OnSomeEvent(string message) {
    X.SomeMethod(message);
  }
  public B(A a) {
    _a = a;
    _a.SomeEvent += OnSomeEvent;
  }
  public void unsubscribe() { _a.SomeEvent -= OnSomeEvent; }
}
```
Here `OnSomeEvent` **implicitly** implements the function signature defined by `Del`, i.e. taking one parameter of type string and returning void.

If I wanted to make it more explicit that I'm implementing the `Del` delegate type, I'd define `OnSomeEvent` as a private delegate member of type `Del` on `B` and assign it a lambda in the constructor.

```csharp
public class B {
  private A _a;
  private Del OnSomeEvent;
  public B(A a) {
    _a = a;
    OnSomeEvent = (string message) => {
      X.SomeMethod(message);
    };
    _a.SomeEvent += OnSomeEvent;
  }
  public void unsubscribe() { _a.SomeEvent -= OnSomeEvent; }
}
```

While this works, it seems too much just for the purpose of making it more explicit that `OnSomeEvent` is supposed to have the same signature as `Del`.

Would it not be fancy to borrow the `public class Something : ISomething` syntax of classes implementing interfaces? I'm imagining something like this:
```csharp
void OnSomeEvent(string message) : Del {
  X.SomeMethod(message);
}
```

If this idea is totally flawed and/or useless please let me know why, so I can gain a better understanding of the language and language design. Thank you for taking to time to read this.

**Comment by HaloFour** [comment]
I'm not sure why that relationship has to be more explicit.  If the method does not match the signature of the delegate then it can't be used to take a delegate of that method, so you already have the explicitness and type-safety that appears to be the goal of this proposal.

If you wanted something more explicit with compile-time checking, you could use something like an analyzer which inspects a custom attribute that specifies the delegate type that the method should match.

**Reply by jnm2** [reply]
E.g.
```cs
[RestrictDelegateTypes(typeof(Del))]
void OnSomeEvent(string message)
{
    X.SomeMethod(message);
}
```

**Reply by jnm2** [reply]
No matter what approach you take, it's very easy to work around though.

```cs
Del del = OnSomeEvent;
Action action = del.Invoke;
```
Or:
```cs
Action action = Shim;
void Shim(string message) => OnSomeEvent(Message);
```

**Comment by TahirAhmadov** [comment]
What I think may be more useful (even though I'm not sure if it's worthwhile) is to somehow define a method using the delegate type as the signature:
```cs
delegate string SomeDel(int a, DateTime b);
// ...
Foo() : SomeDel // syntax I just made up in order to demonstrate
{
  // a and b are implicitly defined, and we have to return string
  return a.ToString() + b.ToString();
}
// or, even using events:
class Bar { event SomeDel SomeEvent; }
// ...
Foo() : Bar.SomeEvent
{
  return a.ToString() + b.ToString();
}
```
Again, I'm not at all sure this is needed, but at least this would be a shorthand for typing out the parameter list.

PS. Really though, this is kind of pointless given that VS already supports some easy actions to auto-generate methods to match events/delegates, so...

**Reply by HaloFour** [reply]
Could always use delegate properties to achieve something similar:

```cs
using System;

var c = new C();
Console.WriteLine(c.Add(2, 3));

public delegate int SomeDel(int x, int y);

public class C {
    public SomeDel Add { get; } = (x, y) => x + y;
}
```
